### ISSUE TITLE
Missing _keras_logits from model.predict(x) leads to difference in loss evaluation

### ISSUE THREAD

**Issue by don-tpanic** [tensorflow/tensorflow#51574]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.4
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.4.1
- Python version: 3.9.2
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): 7.5.0
- CUDA/cuDNN version: 11.2/7.6.5 
- GPU model and memory: GeForce RTX 2080ti, 11019MiB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
`model.predict(x)` does not have attribute `_keras_logits`
**Describe the expected behavior**
`model.predict(x)` and `model(x)` produce identical outputs.
**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```
import numpy as np
import tensorflow as tf
np.random.seed(999)

loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=False)
y_true = [[1, 0, 0]]
inputs = tf.keras.layers.Input(shape=(3,))
outputs = tf.keras.layers.Dense(3, 
            activation='softmax',
            kernel_initializer=tf.keras.initializers.glorot_normal(seed=42))(inputs)
model = tf.keras.models.Model(inputs=inputs, outputs=outputs)

x = np.random.random((1, 3))
y_pred1 = model.predict(x)
y_pred2 = model(x)

# losses are different. One uses logits, the other doesn't.
print('model.predict(x) loss = ', loss_fn(y_true, y_pred1))
print('model(x) loss = ', loss_fn(y_true, y_pred2))  

assert hasattr(y_pred1, '_keras_logits') == hasattr(y_pred2, '_keras_logits')
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by mohantym** [comment]
Hi! @Saduf2019 , Could you please look into this issue , providing [gist](https://colab.research.google.com/gist/mohantym/f9cfabd735f161da34c423a2bab113ff/github_51574.ipynb#scrollTo=cEQWVWqmiy6v) replicated in TF 2.3,2.4,2.5 for your reference. Issue is not occurring in TF 2.3 but afterwards .

**Comment by don-tpanic** [comment]
Thanks for looking into this. Related, it looks like the current way `binary_crossentropy` is written will always use logits so long as there is `_keras_logits` attribute cached during inference and ignore `from_logits=` provided by the user. Is this intended? 

I was referring to this code block inside `def binary_crossentropy(target, output, from_logits=False):`
```
if hasattr(output, '_keras_logits'):
    output = output._keras_logits  # pylint: disable=protected-access
    if from_logits:
      warnings.warn(
          '"`binary_crossentropy` received `from_logits=True`, but the `output`'
          ' argument was produced by a sigmoid or softmax activation and thus '
          'does not represent logits. Was this intended?"')
    from_logits = True
```

**Comment by Abhishek-TyRnT** [comment]
It has been done purposely I guess...... Cause you only convert it to only logits when you are training..... You don't need to logits when you infering

**Comment by don-tpanic** [comment]
> It has been done purposely I guess...... Cause you only convert it to only logits when you are training..... You don't need to logits when you infering

Sorry I didn't mean inference but forward pass to compute the loss. Doesn't it look to you that logits will be used for training even when user set `from_logits=False`?

**Comment by Abhishek-TyRnT** [comment]
If you are using it for training then ....then use model(x).....Cause if you use model.predict function inside a function which is wrapped around @tf.function.... Then it throws an error. model.predict is only used while inferring while model(x) is used when you are training
